### PR TITLE
Handle GlobalKey reparenting inside LayoutBuilder

### LINKED ISSUE #5616: LayoutBuilder is incompatible with GlobalKey reparenting right now
Consider two LayoutBuilders each with a KeyedSubtree child which, in one frame, switch their keys around. (This isn't academic, heroes can trigger something similar during navigation.)

Right now, this will assert because one of the layout builders will try to modify the other layout builder's subtree at layout time when that second one not being laid out (though it is guaranteed that it won't have been laid out yet and is still dirty).

We should probably allow any dirty node to be modified, not just if it's currently being built.

### PR DESCRIPTION
**Pull request by Hixie**
Fixes #5616

### CODE REVIEW

**Comment by Hixie** [comment]
It strikes me that maybe this affects other places that call invokeLayoutCallback. I should audit them.

**Comment by Hixie** [comment]
ok, I made sure this works for those also.

cc @HansMuller for review.

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -1174,6 +1193,10 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
           result = true;
           break;
         }
+        if (owner != null && owner._debugAllowMutationsToDirtySubtrees && node._needsLayout) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/object.dart]: The enclosing function's doc has a typo, returns null should be returns false.

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -880,6 +880,25 @@ class PipelineOwner {
...
 
+  // This flag is used to allow the kinds of mutations performed by GlobalKey
+  // reparenting while a LayoutBuilder is being rebuilt and in so doing tries to
+  // move a node from another LayoutBuilder subtree that hasn't been updated
+  // yet. To set this, call [_enableMutationsToDirtySubtrees], which is called
+  // by [RenderObject.invokeLayoutCallback].
+  bool _debugAllowMutationsToDirtySubtrees = false;
+
+  // See [RenderObject.invokeLayoutCallback].
+  void _enableMutationsToDirtySubtrees(VoidCallback callback) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/object.dart]: I don't understand why this isn't part of invokeLayoutCallback?

**Comment by HansMuller** [comment]
LGTM

File: packages/flutter/lib/src/rendering/object.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: it's on a different class.

File: packages/flutter/lib/src/rendering/object.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: I'm pretty sure it actually returns null in release builds.

**Pull request merged**